pennylane/circuit_drawer/drawable_layers.py

```diff
@@ -0,0 +1,130 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains a helper function to sort operations into layers.
+"""
+
+from .utils import default_wire_map
+
+
+def _recursive_find_layer(checking_layer, op_occupied_wires, occupied_wires_per_layer):
+    """Determine correct layer for operation with ``op_occupied_wires``
+
+    Args:
+        checking_layer (int): function determines if operation fits on this layer
+        op_occupied_wires (set(int)): wires covered the drawn operation.  Includes everything
+            between used wires in a multi-wire gate.
+        occupied_wires_per_layer (list[set[int]]): which wires will already have something drawn
+            on them
+
+    Returns:
+        int: layer to place relevant operation in
+    """
+
+    if occupied_wires_per_layer[checking_layer] & op_occupied_wires:
+        # this layer is occupied, use higher one
+        return checking_layer + 1
+    if checking_layer == 0:
+        # reached first layer, so stop
+        return 0
+    # keep pushing the operation back
+    return _recursive_find_layer(checking_layer - 1, op_occupied_wires, occupied_wires_per_layer)
+
+
+def drawable_layers(ops, wire_map=None):
```

Comment:
The name here could be important. Could this be the "lazy" counterpart to the greedy algo that is being used already in the codebase? A good name could help distinguish how this addition is different to the existing implementation (my understanding is that it's a different approach that is also more performant).